### ISSUE TITLE
FR: `uv build` accepts a dynamic "package version" argument

### ISSUE THREAD

**Issue by charlesnicholson** [astral-sh/uv#8714]
When we build "official" Python distribution package wheels from CI, we version the packages dynamically based on the current git tag. This is surprisingly complicated to do!

The `build` package doesn't support a dynamic version supplied from the command-line: https://build.pypa.io/en/stable/#python--m-build-options

So, we ended up using this thing: https://github.com/pypa/setuptools-scm?tab=readme-ov-file#pyprojecttoml-usage

Our pyproject.toml files look like this:
```
[project]
name = "my_company.package"
dynamic = ["version"]
...
```

And then we have to provide the version as an environment variable to the build:
```
SETUPTOOLS_SCM_PRETEND_VERSION=1.2.3 path/to/venv/bin/python -m build ...
```

It's all very roundabout and contrived; I just want to do this:
```
path/to/venv/bin/python -m build --version=1.2.3 ...
```

I don't know if there's appetite in extending `uv build` to solve this with fewer packages and with a simple command-line, but it would make our lives a lot simpler, and we'd be able to ditch both the `build` and the `setuptools-scm` packages in favor of stock `uv`.

**Comment by zanieb** [comment]
Seems sensible, although we probably need our own build backend to support this? I presume setting the dynamic version is build backend dependent and we're unlikely to tie into their options.

You could also change the version statically, per the workflow in https://github.com/astral-sh/uv/issues/6298, though that's annoying because if you want the committed `pyproject.toml` to be correct you need to do it ahead of time.

As a simple workaround, I set the version to `0.0.0` then do [`sed -i -e "s/0.0.0/${GITHUB_REF#refs/*/}/" pyproject.toml`](https://github.com/astral-sh/packse/blob/70abfe8f64a9746452c02cb514942f879c7eaccc/.github/workflows/release.yaml#L34-L36).

**Comment by charlesnicholson** [comment]
Thanks for the response! Yeah, I landed on the same options you did, I think- in the end I chose not to do in-place modifications of the git `pyproject.toml` files b/c we can sometimes do a "versioned" build locally and we have to remember to git-reset all of the toml files. So for us it's "better" to do this Rube Goldberg silliness :)

**Comment by zanieb** [comment]
There's https://github.com/astral-sh/uv/issues/3957 to track implementation of our own build backend.

**Comment by paveldikov** [comment]
+1 on a UX/ecosystem level. Versioning is such a fundamental build concern and it is so unfortunate that developers have to dance around it with build-backend plugins -- it's super clunky and fragile. I can similarly attest to having burnt a lot of time shimming things with `setuptools-scm` and the like. Frankly, package versioning should have been externalised to begin with.

Architecturally though I'm a -0.5 as this somewhat violates PEP 517. In the current standard, dynamic versioning is fully controlled by the build-backend (as regrettable as this is). And so the only way for the frontend can influence the choice of version, is if the backend is actively willing to be influenced.

I.e. `uv build --version 1.2.3` would _only_ reliably result in a `1.2.3` version if and only if the build-backend is `uv` itself, or is similarly aware of some `uv`-specific build config API. A non-`uv`-aware backend would silently ignore this option. This _may_ be an acceptable compromise in the interim, FWIW.

(IMHO this problem should ultimately be solved at the PEP level)

**Comment by charlesnicholson** [comment]
Agreed, after reading up more on this, a `uv` build backend seems like the only sane place for such an implementation to live.

**Comment by wu-clan** [comment]
pdm does a great job at this：https://pdm-project.org/en/latest/reference/pep621/

**Comment by zanieb** [comment]
@wu-clan I don't see how that's any different than uv? That just looks like a PEP 621 reference guide.

**Comment by wu-clan** [comment]
@zanieb Yes, if you look closely, you'll find https://backend.pdm-project.org/metadata/#dynamic-project-version, which is a very useful feature, especially for libraries

**Comment by vancromy** [comment]
With uv_build now in preview mode and pushed into main it would be amazing to get this FR implemented. We've been using https://github.com/mtkennerly/poetry-dynamic-versioning so far (very similar to setuptools_scm and hatch-vcs) and it would be the last hurdle for us to port over to uv (which is so fast :rocket:) :)

**Comment by snus-kin** [comment]
Not sure if there's a better place to put this comment but now that the uv build backend is stable is this issue tracked elsewhere? I'd like to use it but am stumbling over dynamic versioning, we'd previously used setuptools and a `version.py` file for this. `sed` is a possible workaround but would be nice if it was supported more officially.

**Comment by seandstewart** [comment]
Piling on here - a `--version=` flag would be great and is a general way to solve dynamic versions and is the only thing blocking my organization from adoption of uv's new build backend. 

It would also be very slick to have some sort of extension like https://github.com/ofek/hatch-vcs which could do the work of pulling in the version number from VCS and injecting it into the package metadata directly.

**Comment by zanieb** [comment]
Can ya'll clarify why running `uv version` before `uv build` is insufficient?

**Comment by jzazo** [comment]
I was using [uv-dynamic-versioning](https://github.com/ninoseki/uv-dynamic-versioning) to calculate dev versions when I push to main. When I tag a version manually, then `uv version` would work, but then this has to be committed and locked prior to publishing the package (which I guess CI could do). But the dynamic solution allows to not commit anything while relying on VCS tags and works also for dev versions.

**Comment by seandstewart** [comment]
I've done this many ways. The four main patterns I've come across or used:

Static version artifact, automated versioning:
1. On merge to main, generate new version
2. Update `version` in my toml and/or the `__version__` const in the package ->
3. Commit + tag the change
4. Push new commit + tag to main
5. On tag creation
6. Build distribution and upload
7. Generate Release artifact.

Dynamic version from vcs, automated versioning:
1. On merge to main, generate new version
2. Tag current commit and push
3. Build distribution and upload to registry
4. Generate Release and upload distribution artifacts.

Static version artifact, manual versioning:
1. Create Release artifact (automatically creates your tag).
2. On tag creation
3. Update the version in my toml and/or the `__version__` const in the package.
4. Commit + push the change
5. Build distribution and upload to registry
6. Update Release with distribution artifacts.

Dynamic version artifact, manual versioning:
1. Create Release artifact (automatically creates your tag).
2. On tag creation
3. Build distribution and upload to registry
4. Update Release with distribution artifacts.

Removing a static version from the toml/package reduces the complexity of your CD pipeline. It also means you don't have to juggle permissions and secrets for enabling the automation to push to source control.

Another thing worth noted that is hinted at in the above comment - if you don't have to update the toml, then you don't have to regenerate your lock. This is really useful for code bases which may ship a container image, as it means they don't have to jump through hoops to avoid busting the container build cache just because they updated the version number of the application. Here's an example Dockerfile which demonstrates how one might dance around this issue (pulled from a real production application using poetry - org-specifics redacted):

```Dockerfile
FROM python:3.12 as poetry

WORKDIR /app

ENV POETRY_VERSION="1.8.5"
ENV POETRY_URL="https://install.python-poetry.org"

RUN curl -sSL "${POETRY_URL}" | python3 -

COPY pyproject.toml pyproject.toml
COPY poetry.lock poetry.lock
RUN /root/.local/bin/poetry export --only=main -f requirements.txt -o requirements.txt

FROM python:3.12-slim as dependencies

WORKDIR /app

ENV PIP_DEFAULT_TIMEOUT=100 \
    PIP_DISABLE_PIP_VERSION_CHECK=1 \
    PIP_NO_CACHE_DIR=1 \
    PIP_PREFER_BINARY=1 \
    PIP_NO_INPUT=1

COPY --from=poetry /app/requirements.txt requirements.txt

RUN python -m venv .venv \
    && .venv/bin/pip install -r requirements.txt

FROM python:3.12-slim as main

WORKDIR /app

ENV PYTHONFAULTHANDLER=1 \
    PYTHONHASHSEED=random \
    PYTHONUNBUFFERED=1 \
    PIP_DISABLE_PIP_VERSION_CHECK=1 \
    TZ="UTC"

COPY --from=dependencies /app/.venv .venv
COPY schema schema
COPY src src
COPY pyproject.toml pyproject.toml
RUN .venv/bin/pip install -e .

EXPOSE 8080/tcp
ENTRYPOINT [".venv/bin/myapp"]
CMD ["start"]
```

As you can see, you have to create a dedicated stage just to extract the dependency pins in a way which doesn't depend upon the pyproject.toml until *after* you've installed the dependencies into your build env. If the versioning were dynamic, the Dockerfile could be as little as a single stage.

**Comment by cbrnr** [comment]
I've used the dynamic version feature with hatchling because I want to have the `__version__` attribute in a single place (e.g., `__init__.py`), and I don't want to manually sync it in another place (i.e., `pyproject.toml`).

**Comment by snus-kin** [comment]
> Can ya'll clarify why running `uv version` before `uv build` is insufficient?

Cheers this works better for me I think, doing something like this for the version attribute to use in code (we use this for e.g. logging)

```py
import importlib.metadata

__version__ = importlib.metadata.version("my_module")
```

Worth mentioning that the service we're using for this is pretty lax around versioning anyways, probably this is less appropriate for libraries etc.

**Comment by konstin** [comment]
> I've used the dynamic version feature with hatchling because I want to have the `__version__` attribute in a single place (e.g., `__init__.py`), and I don't want to manually sync it in another place (i.e., `pyproject.toml`).

For this case specifically I strongly recommend the `importlib.metadata` solution that @snus-kin shared. Nowadays, `pyproject.toml` is much better suited as the canonical location of this information, where it can be discovered by tools.

**Comment by my1e5** [comment]
@snus-kin 

> now that the uv build backend is stable is this issue tracked elsewhere?

There is

* https://github.com/astral-sh/uv/issues/14037
* https://github.com/astral-sh/uv/issues/11718

**Comment by charlesnicholson** [comment]
It's a little subtle, but if this is planned to be built I'd greatly prefer a way to customize the version string instead of having uv simply grab the tag straight from git. Sometimes we want to edit or decorate the tag, like removing suffix metadata etc. Being able to pass an arbitrary (legal) version as a command-line arg or environment variable to `uv build` would be much more flexible.

**Comment by Badg** [comment]
+1 to @charlesnicholson -- I just came across an additional use case for dynamic versions, except in this case I really need to be able to specify it directly, and not have it be inferred from a git tag.

**I'd echo the suggestion to allow a dynamic version via an environment variable or commandline flag.** This would allow CI scripts to do whatever git-fu they wanted to grab specific tags (instead of UV needing to add a bunch of special logic for filtering, like PDM does). It would also allow, for example, automatically creating RC builds on pull requests **before** creating a tag. And a whole bunch of other applications.

In other words, I think an environment variable would be a nice MVP escape hatch; worst case, it would give users a way to inject the information into UV while running within a subprocess. And then they could write their own arbitrary code to calculate the version however they wanted.

**Comment by charlesnicholson** [comment]
I gave up on waiting for this- for wheeling, we copy the entire package directory to a staging ground, clean it out, inject dynamic data files that we want bundled up but don't live in the source directory (e.g. native C dlls), and build it. That staging ground gives us a good place to just use our own little toml rewriter to change the version to whatever we want it to be without having to deal with `uv version`, which has an unskippable dependency resolution phase before it will update the toml file.

It's clunkier than uv doing what we want it to here, OTOH it's done and works and we've moved on.

**Comment by Nohac** [comment]
> Can ya'll clarify why running `uv version` before `uv build` is insufficient?

Running `uv version` before `uv build` works and is what we use today, but it doesn’t fit well with our CI workflow. We build and publish many packages from a single workspace, and this approach requires updating each package’s `pyproject.toml`, which is slow and can cause unnecessary lockfile updates. In our setup, package versions aren’t managed separately. The release version is calculated centrally from the workspace major version and the pipeline build number as the minor version. Since that version is already deterministic, we’d prefer to set it at build time instead of rewriting package metadata files for every release.

**Comment by peacefulotter** [comment]
In my case, using a dynamic version and setuptools, `uv version` raises: 

``` 
error: We cannot get or set dynamic project versions in: pyproject.toml

```

A dynamic version is _needed_ for my CI to work...

**Comment by PierreKiwi** [comment]
I use `hatch/hatch-vcs` and namespaced tags for my monorepo.
For instance, in `apps/my-project`, I have:

```toml
[build-system]
requires = ["hatchling", "hatch-vcs"]
build-backend = "hatchling.build"

[project]
name = "my-project"
requires-python = ">=3.13,<3.14"

dynamic = ["version"]

dependencies = [
    "boto3>=1.42,<2",
]

[tool.hatch.build.targets.wheel]
artifacts = ["src/my_project/aws/py.typed"]
packages = ["src/my_project"]

[tool.hatch.version]
fallback-version = "0.0.0"
source = "vcs"
tag-pattern = "^my-project/v(?P<version>[\\d]+\\.[\\d]+\\.[\\d]+.*)$"

[tool.hatch.version.raw-options]
git_describe_command = "git describe --long --tags --match 'my-project/v*'"
relative_to = "pyproject.toml"
root = "../.."
search_parent_directories = true
```

Note: I do not have `__version__.py` file

**Comment by wagenrace** [comment]
If you want a `__version__` in your package you can use this:

```python
from importlib.metadata import version

__version__ = version(__package__)
```

`importlib` is a first-party package.
This means you only have to define you version number in `pyproject.toml` with work around of @zanieb :  `uv version ${GITHUB_REF#refs/*/}`.

**Comment by zanieb** [comment]
You can use `uv version` instead of `sed` now.

**Comment by charlesnicholson** [comment]
@zanieb except, of course, for when you can't :(

https://github.com/astral-sh/uv/issues/15286

**Comment by dhirschfeld** [comment]
> [@zanieb](https://github.com/zanieb) except, of course, for when you can't :(
> 

My CI workflows won't build from a dirty checkout so modifying the `pyproject.toml` is out of the question. I need some way to inject the version for `uv_build` to use. Until then I will stick with `hatch` and `hatch-vcs`.

Hopefully in future `uv_build` will support a dynamic version field injected by an environment variable or otherwise.
